contrib/udfs/src/main/java/org/apache/drill/exec/udfs/NearestDateUtils.java

```diff
@@ -0,0 +1,149 @@
+
+  /**
+   * This function takes a Java LocalDateTime object, and an interval string and returns
+   * the nearest date closets to that time.  For instance, if you specified the date as 2018-05-04 and YEAR, the function
+   * will return 2018-01-01
+   *
+   * @param d        the original datetime before adjustments
+   * @param interval The interval string to deduct from the supplied date
+   * @return the modified LocalDateTime
+   */
+  public final static java.time.LocalDateTime getDate(java.time.LocalDateTime d, String interval) {
+    java.time.LocalDateTime newDate = d;
+    int year = d.getYear();
+    int month = d.getMonth().getValue();
+    int day = d.getDayOfMonth();
+    int hour = d.getHour();
+    int minute = d.getMinute();
+    int second = d.getSecond();
+    TimeInterval adjustmentAmount;
+    try {
+      adjustmentAmount = TimeInterval.valueOf(interval.toUpperCase());
+    } catch (IllegalArgumentException e) {
+      throw new DrillRuntimeException(String.format("[%s] is not a valid time statement. Expecting: %s", interval, Arrays.asList(TimeInterval.values())));
+    }
+    switch (adjustmentAmount) {
+      case YEAR:
+        newDate = LocalDateTime.of(year, 1, 1, 0, 0, 0);
+        break;
+      case QUARTER:
+        newDate = LocalDateTime.of(year, (month / 3) * 3 + 1, 1, 0, 0, 0);
+        break;
+      case MONTH:
+        newDate = LocalDateTime.of(year, month, 1, 0, 0, 0);
+        break;
+      case WEEK_SUNDAY:
+        newDate = newDate.with(TemporalAdjusters.previousOrSame(DayOfWeek.SUNDAY))
+                .truncatedTo(ChronoUnit.DAYS);
+        break;
+      case WEEK_MONDAY:
+        newDate = newDate.with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY))
+                .truncatedTo(ChronoUnit.DAYS);
+        break;
+      case DAY:
+        newDate = LocalDateTime.of(year, month, day, 0, 0, 0);
+        break;
+      case HOUR:
+        newDate = LocalDateTime.of(year, month, day, hour, 0, 0);
+        break;
+      case HALF_HOUR:
+        if (minute >= 30) {
+          minute = 30;
+        } else {
+          minute = 0;
+        }
+        newDate = LocalDateTime.of(year, month, day, hour, minute, 0);
+        break;
+      case QUARTER_HOUR:
+        if (minute >= 45) {
+          minute = 45;
+        } else if (minute >= 30) {
+          minute = 30;
+        } else if (minute >= 15) {
+          minute = 15;
+        } else {
+          minute = 0;
+        }
+        newDate = LocalDateTime.of(year, month, day, hour, minute, 0);
+        break;
+      case MINUTE:
+        newDate = LocalDateTime.of(year, month, day, hour, minute, 0);
+        break;
+      case HALF_MINUTE:
+        if (second >= 30) {
+          second = 30;
+        } else {
+          second = 0;
+        }
+        newDate = LocalDateTime.of(year, month, day, hour, minute, second);
+        break;
+      case QUARTER_MINUTE:
+        if (second >= 45) {
```

Comment:
This if-else and similar constructs above may be refactored using method, like: 

```java
  private static int roundDownTo(int value, int... roundingCandidates) {
    for (int candidate : roundingCandidates) {
      if (value >= candidate) return candidate;
    }
    return roundingCandidates[roundingCandidates.length - 1];
  }
```
Then code here and above will look like: 
```java
minute = roundDownTo(minute, 45, 30, 15, 0);
```

Reply:
What's the advantage of this? Wouldn't this add additional overhead from the function calls and loops?

Reply:
That's a good point. Advantage is that such method greatly reduces the ```if-else``` code, but you're right that varargs will create new int array for every method call. So I have to agree that this suggestion wasn't a good idea, sorry for distraction.